### PR TITLE
Merge candidates

### PR DESCRIPTION
**Pull request by jack13only**
image

### CODE REVIEW

**Review by AlexKolonitskyMifort — CHANGES_REQUESTED** [review]

File: src/app/app.component.html [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <app-navbar></app-navbar>
-<app-actions-bar></app-actions-bar>
+<!-- <app-actions-bar></app-actions-bar> -->
```
[the remark this answers is not in the document]
Comment (jack13only) [reply src/app/app.component.html]: done

File: src/app/components/systems/actions-bar-merge/actions-bar-merge.component.html [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<div class="actions-wrapper">
+  <button mat-raised-button class="mb-20" (click)="back()">
```
[the remark this answers is not in the document]
Comment (jack13only) [reply src/app/components/systems/actions-bar-merge/actions-bar-merge.component.html]: done

File: src/app/components/systems/actions-bar-merge/actions-bar-merge.component.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+@Component({
+  selector: 'app-actions-bar-merge',
+  templateUrl: './actions-bar-merge.component.html',
+  styleUrls: ['./actions-bar-merge.component.scss'],
+})
+export class ActionsBarMergeComponent {
+  constructor(public mergeService: MergeService, private router: Router) {}
+
+  back(): void {
```
[the remark this answers is not in the document]
Comment (jack13only) [reply src/app/components/systems/actions-bar-merge/actions-bar-merge.component.ts]: done

File: src/app/components/systems/actions-bar/actions-bar.component.html [hunk 4]
```diff
@@ -4,4 +4,4 @@
       <mat-icon>group_add</mat-icon><span class="ml-10">New candidate</span>
     </button>
   </div>
-</mat-toolbar>
+</mat-toolbar>
```
Comment (AlexKolonitskyMifort) [inline src/app/components/systems/actions-bar/actions-bar.component.html]: revert

File: angular.json [hunk 5]
```diff
@@ -105,6 +105,7 @@
   },
   "defaultProject": "hr-automation-ui",
   "cli": {
+    "analytics": "26435a0f-7bab-4839-a314-5b71491ee461",
```
Comment (AlexKolonitskyMifort) [inline angular.json]: что это и зачем оно тут?)

File: src/app/pages/merge-page/actions-bar-merge/actions-bar-merge.component.scss [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    display: flex;
+    justify-content: right;
+    padding: 10px 15px 0 30px;
+}
+  
+  .buttons-wrapper {
+    text-align: right;
+  }
+  
+  .attachment {
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/actions-bar-merge/actions-bar-merge.component.scss]: why you need this style?

File: src/app/pages/merge-page/merge-candidate/merge-candidate.component.scss [hunk 7]
```diff
@@ -0,0 +1,40 @@
+$primary-color: rgb(55, 111, 252);
+$dark-back-color: rgb(10, 10, 10);
+$light-back-color: rgb(121, 121, 121);
+$error-color: rgb(121, 1, 1);
+$box-shadow: 0px 2px 1px -1px rgba(0, 0, 0, 20%), 0px 1px 1px 0px rgba(0, 0, 0, 14%), 0px 1px 3px 0px rgba(0, 0, 0, 12%);
+$height: 12vh;
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-candidate/merge-candidate.component.scss]: all variables should be moved to common styles

File: src/app/pages/merge-page/merge-cell/merge-cell.component.scss [hunk 8]
```diff
@@ -0,0 +1,35 @@
+$primary-color: rgb(55, 111, 252);
+$dark-back-color: rgb(10, 10, 10);
+$light-back-color: rgb(121, 121, 121);
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-cell/merge-cell.component.scss]: move to common styles

File: src/app/pages/merge-page/merge-cell/merge-cell.component.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  constructor(public mergeService: MergeService) {}
+
+  @Input() value?: MergeCandidateAttribute[];
+
+  @Input() editable!: boolean;
+
+  @Output() valueChange = new EventEmitter<boolean>();
+
+  onCheckboxChange(item: MergeCandidateAttribute) {
+    // eslint-disable-next-line no-param-reassign
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-cell/merge-cell.component.ts]: I guess you can remove supression

File: src/app/pages/merge-page/merge-page.component.scss [hunk 10]
```diff
@@ -0,0 +1,60 @@
+$primary-color: rgb(55, 111, 252);
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.scss]: move to common styles

File: src/app/pages/merge-page/merge-page.component.scss [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    gap: 10px; 
+    overflow-x: auto;  
+    
+    .desc-candidate-container {    
+      min-width: 300px;
+      overflow-y: hidden;
+      padding-bottom: 5px;
+    }
+    
+    .candidates-container-titles {
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.scss]: is it used?

File: src/app/pages/merge-page/merge-page.component.html [hunk 12]
```diff
@@ -0,0 +1,52 @@
+<app-actions-bar-merge></app-actions-bar-merge>
+<mat-progress-bar
+  mode="indeterminate"
+  color="primary"
+  *ngIf="pageState.loading"
+></mat-progress-bar>
+<ng-container *ngIf="!pageState.loading">
+  <div *ngIf="candidatesMatrixNotEmpty; else noCandidatesId" class="candidate-merge-container">
+    <div class="desc-titles-container">
+      <div class="desc-title"></div>
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.html:10]: why we need this div element?

File: src/app/pages/merge-page/merge-page.component.html [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+          <span>&nbsp;</span>        
+          {{attributeType.name}}
+        </div>
+      </ng-container>
+  </div>
+
+    <div class="candidates-container">
+      <ng-container *ngFor="let candidate of candidates">
+        <app-merge-candidate
+          class="desc-candidate-container"
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.html]: is it possible to move class with styles inside app-merge-candidate

File: src/app/pages/merge-page/merge-page.component.scss [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+$light-back-color: rgb(121, 121, 121);
+$error-color: rgb(121, 1, 1);
+$box-shadow: 0px 2px 1px -1px rgba(0, 0, 0, 20%), 0px 1px 1px 0px rgba(0, 0, 0, 14%), 0px 1px 3px 0px rgba(0, 0, 0, 12%);
+$height: 12vh;
+
+.candidate-merge-container {
+  margin: 20px 100px;
+  padding-bottom: 10px;
+  height: auto;
+  font-family: Roboto, "Helvetica Neue", sans-serif;
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.scss]: let's use common styles in all application

File: src/app/pages/merge-page/merge-page.component.ts [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  finalResult(): MergeCandidate {
+    return this.candidates.reduce((res, candidate) => {
+      res.id = 'Results';
+      res.attributes = (res.attributes || []).concat(
+        candidate.attributes.filter((a) => a.selected)
+      );
+      res.attributesMap = new Map<string, MergeCandidateAttribute[]>(
+        Object.entries(
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.ts:60]: use [mapValues ](https://lodash.com/docs/4.17.15#mapValues) from lodash

File: src/app/pages/merge-page/merge-page.component.ts [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+  constructor(public mergeService: MergeService) {
+    this.candidatesMatrix = this.mergeService
+      .getCandidates()
+      .pipe(map((candidates) => new MergeCandidates(candidates)));
+  }
+
+  ngOnInit(): void {
+    this.pageState.startLoading();
+    this.candidatesMatrix.subscribe((matrix) => {
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.ts]: add error hadling

File: src/app/pages/merge-page/merge-page.component.ts [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+      if (matrix) {
+        this.pageState.finishLoading();
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.ts:48]: move [finalize](https://www.learnrxjs.io/learn-rxjs/operators/utility/finalize)

File: src/app/pages/merge-page/merge-page.component.ts (hunk 16 above)
Comment (AlexKolonitskyMifort) [reply src/app/pages/merge-page/merge-page.component.ts]: handle unsubscribe

File: src/app/pages/merge-page/merge.service.ts [hunk 18]
```diff
@@ -0,0 +1,70 @@
+/* eslint-disable no-console */
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge.service.ts]: let's remove supression

File: src/app/services/candidates.service.ts [hunk 19]
```diff
@@ -45,9 +45,11 @@ export class CandidatesService {
     );
   }
 
-  public getCandidateById(id: string): any {
+  public getCandidateById(id: string): Observable<Candidate> {
+    // eslint-disable-next-line no-console
+    console.log('CandidatesService.getCandidateById', id);
```
Comment (AlexKolonitskyMifort) [inline src/app/services/candidates.service.ts]: let's remove console.log()

File: src/app/pages/merge-page/merge-page.component.ts [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        []
+      );
+  }
+
+  getAllAttributeTypesFrom(): CandidateAttributeType[] {
+    return this.allAttributeTypes;
+  }
+
+  public selectAttribute(
+    candidate: MergeCandidate,
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.ts]: remove candidate

File: src/app/pages/merge-page/merge-page.component.ts [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+import { Component, OnInit } from '@angular/core';
+import { map, Observable } from 'rxjs';
+import * as _ from 'lodash';
+import { MergeService } from '@pages/merge-page/merge.service';
+import { PageState } from '@src/app/utils/pageState';
+import { CandidateAttributeType } from '@src/app/models/candidates';
+import { MergeCandidate } from '@pages/merge-page/view-model/MergeCandidate';
+import { MergeCandidateAttribute } from '@pages/merge-page/view-model/MergeCandidateAttribute';
+
+class MergeCandidates {
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.ts]: move this class to separate file

File: src/app/pages/merge-page/merge-page.component.ts [hunk 22]
```diff
@@ -0,0 +1,105 @@
+/* eslint-disable no-console */
+// eslint-disable-next-line max-classes-per-file
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.ts]: remove eslint supressoins

File: src/app/pages/merge-page/merge-page.component.ts [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+  private readonly candidates: MergeCandidate[];
+
+  private readonly allAttributeTypes: CandidateAttributeType[];
+
+  constructor(candidates: MergeCandidate[]) {
+    this.candidates = candidates;
+    this.allAttributeTypes = _(candidates)
+      .flatMap((c) => c.attributes.map((a) => a.attributeTypes))
+      .uniqBy('id')
+      .reduce(
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.ts]: is it possible to simplify it?

File: src/app/pages/merge-page/merge-page.component.scss [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+  gap: 10px;
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/merge-page.component.scss]: move gap size into variable

**Review by AlexKolonitskyMifort — CHANGES_REQUESTED** [review]

File: src/app/components/systems/actions-bar/actions-bar.component.html (hunk 4 above)
Comment (jack13only) [reply src/app/components/systems/actions-bar/actions-bar.component.html]: done

File: src/app/pages/merge-page/merge-candidate/merge-candidate.component.scss (hunk 7 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-candidate/merge-candidate.component.scss]: done

File: src/app/pages/merge-page/merge-cell/merge-cell.component.scss (hunk 8 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-cell/merge-cell.component.scss]: done

File: src/app/pages/merge-page/merge-cell/merge-cell.component.ts (hunk 9 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-cell/merge-cell.component.ts]: done

File: src/app/pages/merge-page/merge-page.component.scss (hunk 10 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.scss]: done

File: src/app/pages/merge-page/merge-page.component.scss (hunk 11 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.scss]: done

File: src/app/pages/merge-page/merge-page.component.html (hunk 12 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.html:10]: Empty element at the beginning of the table

File: src/app/pages/merge-page/merge-page.component.html (hunk 13 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.html]: done

File: src/app/pages/merge-page/merge-page.component.scss (hunk 14 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.scss]: done

File: src/app/pages/merge-page/merge-page.component.ts (hunk 16 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.ts]: done

File: src/app/pages/merge-page/merge.service.ts (hunk 18 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge.service.ts]: done

File: src/app/services/candidates.service.ts (hunk 19 above)
Comment (jack13only) [reply src/app/services/candidates.service.ts]: done

File: src/app/pages/merge-page/merge-page.component.ts (hunk 20 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.ts]: done

File: src/app/pages/merge-page/merge-page.component.ts (hunk 21 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.ts]: done

File: src/app/pages/merge-page/merge-page.component.ts (hunk 22 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.ts]: done

File: src/app/pages/merge-page/merge-page.component.scss (hunk 24 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.scss]: done

File: src/app/components/systems/actions-bar/actions-bar.component.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (jack13only) [reply src/app/components/systems/actions-bar/actions-bar.component.html]: done

File: src/app/pages/merge-page/actions-bar-merge/actions-bar-merge.component.scss (hunk 6 above)
Comment (jack13only) [reply src/app/pages/merge-page/actions-bar-merge/actions-bar-merge.component.scss]: done

File: src/app/pages/merge-page/merge-page.component.ts (hunk 15 above)
Comment (jack13only) [reply src/app/pages/merge-page/merge-page.component.ts:60]: done

File: src/app/pages/merge-page/actions-bar-merge/actions-bar-merge.component.html [hunk 25]
```diff
@@ -0,0 +1,10 @@
+<div class="actions-wrapper">
+  <div class="gap-10">
+    <div class="col-12 col-10-sm buttons-wrapper">
```
Comment (AlexKolonitskyMifort) [inline src/app/pages/merge-page/actions-bar-merge/actions-bar-merge.component.html:3]: what is class buttons-wrapper?

File: src/app/pages/merge-page/merge-page.component.ts (hunk 17 above)
Comment (AlexKolonitskyMifort) [reply src/app/pages/merge-page/merge-page.component.ts:48]: add error handling

**Review by AlexKolonitskyMifort — APPROVED** [review]

**Pull request merged**